### DISCUSSION TITLE
Potential caveats when using Opti stack?

### DISCUSSION THREAD

**Discussion by itbellix** [casadi/casadi#3259] [Q&A] [answered]
Dear Casadi team,
I have used Casadi about 4 years ago for implementing a NLP during my masters thesis. Now I am coming back to it, and I see that with newer releases of the package the `Opti()`  methods have been introduced. I read a some posts on the old Google forum, and got the sense that there might be some differences in terms of functionality of the `Opti()` stack versus the way in which I was used to implementing an NLP (the "conventional" method, working directly with `SX` and `MX` variables).

Is there anything for which the conventional method would still be recommended?

I am thinking specifically about computational efficiency (I imagine that this stack would be built on top of the same functions, so could have some sort of overhead?) and simplicity of implementing direct collocation methods when solving the MPC problems.
I appreciate your time, and wish you a nice day.

Best regards,
Italo

**Comment by jgillis** [comment, marked as answer]
* Opti does not support lbx/ubx, i.e. simple bounds on decision variables. However since 3.6 introduced 'detect_simple_bounds' option, this is not necessarily an issue.
 * There is overhead in using Opti set_initial, set_value, value. This overhead can be removed by making use of 'to_function'
 * Modeling in Opti is always in MX not SX. You can still use "expand" option to get an SX implementation after the modeling though.
 
 If you are fluent in CasADi, you can almost always obtain equal runtime efficiency in Opti versus nlpsol.
I hope you will enjoy the ride again after 4 years

**Comment by itbellix** [comment]
Hi @jgillis , thank you!
- I am not sure I understand fully your first point: do you mean that using Opti you cannot impose bounds on the optimization variables (as, for example, would be needed in the [`direct_collocation.py` example](https://github.com/casadi/casadi/blob/main/docs/examples/python/direct_collocation.py#L118-L119)?
  Looking at [`direct_collocation_opti.m`](https://github.com/casadi/casadi/blob/main/docs/examples/matlab/direct_collocation_opti.m#L80C4-L80C38) I think should be able to, so I am not sure I am interpreting correctly what you mean when you say that
  > Opti does support lbx/ubx

**Reply by jgillis** [reply]
No bounds on decision variables indeed. Everything is interpreted as lbg<= g(x,p) <= ubg

**Reply by itbellix** [reply]
Thank you, then I understand that users just need to reformulate their bounds on the decision variables in another way, but there is no restriction in terms of the type of problems that can be treated.
So, to summarize, it looks to me that Opti should be the way to go and there is no reason to nostalgically keep using `MX` variables directly!

**Reply by jgillis** [reply]
I will not dispute that conclusion. We have made a decision years ago to keep Opti low profile, in order to avoid competing with modeling libraries built on top of CasADi.

**Comment by itbellix** [comment]
I report here the results of speed comparison hoping they can be useful to someone.

I have implemented the same optimization problem both in the conventional way using `MX` variables directly and making use of the `Opti` stack (using direct collocation, in Python). I chose just a simple example, in which the state represents position and velocity of a point mass on a 2D plane, and the input is a 2D force that can be applied to drive the point-mass towards a desired goal. Effectively, I have an LTI system of the type 
x_dot = Ax+Bu 
where x is 4x1 and u is 2x1

I have set up and solved the same NLP 5000 times, and recorded the average time needed by to create and solve the problem (`total time`) and the time needed by the solver to solve the problem (`time to solve`).

From my results, it seems that using `Opti` is slower on average (by a small margin):

| method | `total time` [s] | `time to solve` [s] |
| --- | --- | --- |
| conventional | 0.093 | 0.031 |
| Opti stack | 0.102 | 0.032 |
